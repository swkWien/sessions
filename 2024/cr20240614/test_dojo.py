"""
Guiding test vs test den wir recyclen?
* Guiding test ist gut aber verlangt viel auf einmal
* Alte Tests sollten stehen bleiben k√∂nnen wie sie sind
Wann testen wir zu viel?

Approval Schritt forciert Kontextwechsel und das ist hilfreich, wir
sehen nur mehr die Daten und nicht mehr den Code
Es gibt eine kleine Pause

TODO
Problem mit der Darstellung: Koordinaten/Gitter falsch?
Alle Animationen in einem Fenster?
"""


from approvaltests import verify_as_json, verify
import attrs
import cattrs
import pytest

@attrs.define
class Coordinate:
    x: int
    y: int

@attrs.define
class Square:
    coordinate: Coordinate
    color: int = 1


@attrs.define
class Ant:
    coordinate: Coordinate = Coordinate(0, 0)
    orientation: int = 0

    def turn_right_and_move(self):
        moved_coordinate = Coordinate(1, 0) if self.orientation == 0 else Coordinate(0, 1)
        moved_orientation = self.orientation-90
        return Ant(moved_coordinate, orientation=moved_orientation)

    def turn_left_and_move(self):
        moved_coordinate = Coordinate(-1, 0)
        moved_orientation = self.orientation+90
        return Ant(moved_coordinate, orientation=moved_orientation)

    def move(self, color):
        if color == 0:
            moved_ant = self.turn_right_and_move()
        else:
            moved_ant = self.turn_left_and_move()
        return moved_ant


@attrs.define
class State:
    squares: list[Square]
    ant: Ant = Ant()

    def transform(self):
        ant_is_on_white_square = True
        for square in self.squares:
            if square.coordinate == self.ant.coordinate:
                ant_is_on_white_square = square.color == 0
                break

        if ant_is_on_white_square:
            color_at_ant = 0
            squares = self.squares + [Square(self.ant.coordinate, color=1)]
        else:
            color_at_ant = 1
            squares = []

        moved_ant = self.ant.move(color_at_ant)
        return State(squares=squares, ant=moved_ant)


@attrs.define
class Config:
    title: str
    figsize: int = 5
    gridsize: int = 5


def test_simulation_with_one_square():

    config = Config("Simulation")
    config_unstructured = cattrs.unstructure(config)

    coordinate = Coordinate(0, 0)
    square = Square(coordinate)
    state = State([square])

    state_unstructured = cattrs.unstructure(state)

    simulation_data = {
        "config": config_unstructured,
        "states": [
            state_unstructured
        ],
    }

    verify_as_json(simulation_data)


def test_simulation_with_ant():

    config = Config("Simulation")
    config_unstructured = cattrs.unstructure(config)

    coordinate = Coordinate(0, 0)
    ant = Ant(coordinate, orientation=0)
    states = [State(squares=[], ant=ant)]
    states_unstructured = cattrs.unstructure(states)

    simulation_data = {
        "config": config_unstructured,
        "states": states_unstructured,
    }

    verify_as_json(simulation_data)


def test_simulation_with_moving_ant():

    config = Config("Simulation")
    config_unstructured = cattrs.unstructure(config)

    starting_position = Coordinate(0, 0)
    state_initial = State(squares=[], ant=Ant(starting_position, orientation=0))

    state_transformed = state_initial.transform()

    states = [
        state_initial,
        state_transformed
    ]
    states_unstructured = cattrs.unstructure(states)

    simulation_data = {
        "config": config_unstructured,
        "states": states_unstructured,
    }

    verify_as_json(simulation_data)

def test_simulation_with_ant_facing_left():

    config = Config("Simulation")
    config_unstructured = cattrs.unstructure(config)

    starting_position = Coordinate(0, 0)
    state_initial = State(squares=[], ant=Ant(starting_position, orientation=90))

    state_transformed = state_initial.transform()

    states = [
        state_initial,
        state_transformed
    ]
    states_unstructured = cattrs.unstructure(states)

    simulation_data = {
        "config": config_unstructured,
        "states": states_unstructured,
    }

    verify_as_json(simulation_data)


def test_simulation_with_ant_starts_from_colored_square():

    config = Config("Simulation")
    config_unstructured = cattrs.unstructure(config)

    starting_position = Coordinate(0, 0)
    state_initial = State(squares=[Square(starting_position, color=1)], ant=Ant(starting_position, orientation=0))

    state_transformed = state_initial.transform()

    states = [
        state_initial,
        state_transformed
    ]
    states_unstructured = cattrs.unstructure(states)

    simulation_data = {
        "config": config_unstructured,
        "states": states_unstructured,
    }

    verify_as_json(simulation_data)


def test_simulation_with_ant_starts_from_white_square_next_to_colored_square():

    config = Config("Simulation")
    config_unstructured = cattrs.unstructure(config)

    starting_position = Coordinate(0, 0)
    state_initial = State(squares=[Square(Coordinate(1, 1), color=1)], ant=Ant(starting_position, orientation=0))

    state_transformed = state_initial.transform()

    states = [
        state_initial,
        state_transformed
    ]
    states_unstructured = cattrs.unstructure(states)

    simulation_data = {
        "config": config_unstructured,
        "states": states_unstructured,
    }

    verify_as_json(simulation_data)