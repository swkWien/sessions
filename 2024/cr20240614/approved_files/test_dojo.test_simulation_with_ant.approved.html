
<link rel="stylesheet"
href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
<script language="javascript">
  function isInternetExplorer() {
    ua = navigator.userAgent;
    /* MSIE used to detect old browsers and Trident used to newer ones*/
    return ua.indexOf("MSIE ") > -1 || ua.indexOf("Trident/") > -1;
  }

  /* Define the Animation class */
  function Animation(frames, img_id, slider_id, interval, loop_select_id){
    this.img_id = img_id;
    this.slider_id = slider_id;
    this.loop_select_id = loop_select_id;
    this.interval = interval;
    this.current_frame = 0;
    this.direction = 0;
    this.timer = null;
    this.frames = new Array(frames.length);

    for (var i=0; i<frames.length; i++)
    {
     this.frames[i] = new Image();
     this.frames[i].src = frames[i];
    }
    var slider = document.getElementById(this.slider_id);
    slider.max = this.frames.length - 1;
    if (isInternetExplorer()) {
        // switch from oninput to onchange because IE <= 11 does not conform
        // with W3C specification. It ignores oninput and onchange behaves
        // like oninput. In contrast, Microsoft Edge behaves correctly.
        slider.setAttribute('onchange', slider.getAttribute('oninput'));
        slider.setAttribute('oninput', null);
    }
    this.set_frame(this.current_frame);
  }

  Animation.prototype.get_loop_state = function(){
    var button_group = document[this.loop_select_id].state;
    for (var i = 0; i < button_group.length; i++) {
        var button = button_group[i];
        if (button.checked) {
            return button.value;
        }
    }
    return undefined;
  }

  Animation.prototype.set_frame = function(frame){
    this.current_frame = frame;
    document.getElementById(this.img_id).src =
            this.frames[this.current_frame].src;
    document.getElementById(this.slider_id).value = this.current_frame;
  }

  Animation.prototype.next_frame = function()
  {
    this.set_frame(Math.min(this.frames.length - 1, this.current_frame + 1));
  }

  Animation.prototype.previous_frame = function()
  {
    this.set_frame(Math.max(0, this.current_frame - 1));
  }

  Animation.prototype.first_frame = function()
  {
    this.set_frame(0);
  }

  Animation.prototype.last_frame = function()
  {
    this.set_frame(this.frames.length - 1);
  }

  Animation.prototype.slower = function()
  {
    this.interval /= 0.7;
    if(this.direction > 0){this.play_animation();}
    else if(this.direction < 0){this.reverse_animation();}
  }

  Animation.prototype.faster = function()
  {
    this.interval *= 0.7;
    if(this.direction > 0){this.play_animation();}
    else if(this.direction < 0){this.reverse_animation();}
  }

  Animation.prototype.anim_step_forward = function()
  {
    this.current_frame += 1;
    if(this.current_frame < this.frames.length){
      this.set_frame(this.current_frame);
    }else{
      var loop_state = this.get_loop_state();
      if(loop_state == "loop"){
        this.first_frame();
      }else if(loop_state == "reflect"){
        this.last_frame();
        this.reverse_animation();
      }else{
        this.pause_animation();
        this.last_frame();
      }
    }
  }

  Animation.prototype.anim_step_reverse = function()
  {
    this.current_frame -= 1;
    if(this.current_frame >= 0){
      this.set_frame(this.current_frame);
    }else{
      var loop_state = this.get_loop_state();
      if(loop_state == "loop"){
        this.last_frame();
      }else if(loop_state == "reflect"){
        this.first_frame();
        this.play_animation();
      }else{
        this.pause_animation();
        this.first_frame();
      }
    }
  }

  Animation.prototype.pause_animation = function()
  {
    this.direction = 0;
    if (this.timer){
      clearInterval(this.timer);
      this.timer = null;
    }
  }

  Animation.prototype.play_animation = function()
  {
    this.pause_animation();
    this.direction = 1;
    var t = this;
    if (!this.timer) this.timer = setInterval(function() {
        t.anim_step_forward();
    }, this.interval);
  }

  Animation.prototype.reverse_animation = function()
  {
    this.pause_animation();
    this.direction = -1;
    var t = this;
    if (!this.timer) this.timer = setInterval(function() {
        t.anim_step_reverse();
    }, this.interval);
  }
</script>

<style>
.animation {
    display: inline-block;
    text-align: center;
}
input[type=range].anim-slider {
    width: 374px;
    margin-left: auto;
    margin-right: auto;
}
.anim-buttons {
    margin: 8px 0px;
}
.anim-buttons button {
    padding: 0;
    width: 36px;
}
.anim-state label {
    margin-right: 8px;
}
.anim-state input {
    margin: 0;
    vertical-align: middle;
}
</style>

<div class="animation">
  <img id="_anim_img3b9ce7f55fc64075901dd4cee46c8c5b">
  <div class="anim-controls">
    <input id="_anim_slider3b9ce7f55fc64075901dd4cee46c8c5b" type="range" class="anim-slider"
           name="points" min="0" max="1" step="1" value="0"
           oninput="anim3b9ce7f55fc64075901dd4cee46c8c5b.set_frame(parseInt(this.value));">
    <div class="anim-buttons">
      <button title="Decrease speed" aria-label="Decrease speed" onclick="anim3b9ce7f55fc64075901dd4cee46c8c5b.slower()">
          <i class="fa fa-minus"></i></button>
      <button title="First frame" aria-label="First frame" onclick="anim3b9ce7f55fc64075901dd4cee46c8c5b.first_frame()">
        <i class="fa fa-fast-backward"></i></button>
      <button title="Previous frame" aria-label="Previous frame" onclick="anim3b9ce7f55fc64075901dd4cee46c8c5b.previous_frame()">
          <i class="fa fa-step-backward"></i></button>
      <button title="Play backwards" aria-label="Play backwards" onclick="anim3b9ce7f55fc64075901dd4cee46c8c5b.reverse_animation()">
          <i class="fa fa-play fa-flip-horizontal"></i></button>
      <button title="Pause" aria-label="Pause" onclick="anim3b9ce7f55fc64075901dd4cee46c8c5b.pause_animation()">
          <i class="fa fa-pause"></i></button>
      <button title="Play" aria-label="Play" onclick="anim3b9ce7f55fc64075901dd4cee46c8c5b.play_animation()">
          <i class="fa fa-play"></i></button>
      <button title="Next frame" aria-label="Next frame" onclick="anim3b9ce7f55fc64075901dd4cee46c8c5b.next_frame()">
          <i class="fa fa-step-forward"></i></button>
      <button title="Last frame" aria-label="Last frame" onclick="anim3b9ce7f55fc64075901dd4cee46c8c5b.last_frame()">
          <i class="fa fa-fast-forward"></i></button>
      <button title="Increase speed" aria-label="Increase speed" onclick="anim3b9ce7f55fc64075901dd4cee46c8c5b.faster()">
          <i class="fa fa-plus"></i></button>
    </div>
    <form title="Repetition mode" aria-label="Repetition mode" action="#n" name="_anim_loop_select3b9ce7f55fc64075901dd4cee46c8c5b"
          class="anim-state">
      <input type="radio" name="state" value="once" id="_anim_radio1_3b9ce7f55fc64075901dd4cee46c8c5b"
             >
      <label for="_anim_radio1_3b9ce7f55fc64075901dd4cee46c8c5b">Once</label>
      <input type="radio" name="state" value="loop" id="_anim_radio2_3b9ce7f55fc64075901dd4cee46c8c5b"
             checked>
      <label for="_anim_radio2_3b9ce7f55fc64075901dd4cee46c8c5b">Loop</label>
      <input type="radio" name="state" value="reflect" id="_anim_radio3_3b9ce7f55fc64075901dd4cee46c8c5b"
             >
      <label for="_anim_radio3_3b9ce7f55fc64075901dd4cee46c8c5b">Reflect</label>
    </form>
  </div>
</div>


<script language="javascript">
  /* Instantiate the Animation class. */
  /* The IDs given should match those used in the template above. */
  (function() {
    var img_id = "_anim_img3b9ce7f55fc64075901dd4cee46c8c5b";
    var slider_id = "_anim_slider3b9ce7f55fc64075901dd4cee46c8c5b";
    var loop_select_id = "_anim_loop_select3b9ce7f55fc64075901dd4cee46c8c5b";
    var frames = new Array(1);
    
  frames[0] = "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAfQAAAH0CAYAAADL1t+KAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90\
bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9h\
AAAPYQGoP6dpAAATSUlEQVR4nO3dfazWdf3H8dfFjcDEW2A2bHrIEnWi5qS8aYE3zYVh2eambUw3\
2/qjUoMUKAmtTeWYiv7SP0yNlnNZs5ulTUqQbhwjnYkVGnmDrqQETASDBLx+f1zzuCNICudwnfM+\
j8c/cH3PdX2/7+uPw5PP93zP92o0m81mAIB+bVC7BwAAdp+gA0ABgg4ABQg6ABQg6ABQgKADQAGC\
DgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoA\
FCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCA\
oANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg69IJly5blnHPOySGHHJJhw4bloIMOykknnZQZM2Z0\
PWfy5MmZPHlyW+ZbsmRJGo1GlixZ8p5fu2LFilx55ZVZtWrVdl+78MIL09HRsdvzAe+doEMPu//+\
+3PyySfn1VdfTWdnZ371q1/lpptuyimnnJJ77rmn63m33nprbr311jZOumtWrFiRq666aodBnzNn\
Tn7605/u+aGADGn3AFBNZ2dnxo0bl4ULF2bIkLe+xc4777x0dnZ2PT7qqKPaMV6vOuyww9o9AgxY\
VujQw9atW5fRo0d3i/mbBg1661vu7afcV61alUajkeuuuy7z5s1LR0dHRowYkcmTJ2flypXZsmVL\
Zs2albFjx2a//fbLOeeck5deeqnb/huNRq688srtjtvR0ZELL7xwp3M/+uijOe+887qO29HRkfPP\
Pz/PP/9813MWLFiQc889N0ly6qmnptFopNFoZMGCBUl2fMp98+bNmT17dsaNG5e99torBx98cL74\
xS/mlVde2W7GT33qU3nggQdy/PHHZ8SIETniiCNy55137nRuoEXQoYeddNJJWbZsWS6++OIsW7Ys\
W7ZseU+vv+WWW/Lwww/nlltuye23356nnnoqU6dOzUUXXZQ1a9bkzjvvTGdnZx588MF8/vOf77G5\
V61alfHjx2f+/PlZuHBh5s2bl9WrV2fixIlZu3ZtkuSss87K1Vdf3TXn0qVLs3Tp0px11lk73Gez\
2cxnPvOZfPvb3860adNy//33Z/r06fn+97+f0047Lf/973+7PX/58uWZMWNGvvKVr+TnP/95jjnm\
mFx00UX57W9/22PvE8pqAj1q7dq1zY997GPNJM0kzaFDhzZPPvnk5jXXXNPcsGFD1/MmTZrUnDRp\
Utfj5557rpmkeeyxxza3bdvWtX3+/PnNJM2zzz6723EuvfTSZpLm+vXru7Ylac6dO3e7mQ499NDm\
BRdc0PX4oYceaiZpPvTQQ+/4PrZu3drcuHFjc++9927edNNNXdt//OMfv+NrL7jgguahhx7a9fiB\
Bx5oJml2dnZ2e94999zTTNK87bbbus04fPjw5vPPP9+1bdOmTc0DDzyw+YUvfOEd5wRarNChh40a\
NSq/+93v8sgjj+Taa6/Npz/96axcuTKzZ8/OhAkTula772TKlCndTs0feeSRSbLdKvjN7S+88EKP\
zL1x48bMnDkzH/zgBzNkyJAMGTIkI0eOzGuvvZYnn3xyl/a5ePHiJNnudP+5556bvffeO4sWLeq2\
/bjjjsshhxzS9Xj48OE5/PDDu532B3bMRXHQS0444YSccMIJSZItW7Zk5syZufHGG9PZ2dnt4ri3\
O/DAA7s93muvvXa6ffPmzT0y7+c+97ksWrQoc+bMycSJE7Pvvvum0WhkypQp2bRp0y7tc926dRky\
ZEjGjBnTbXuj0cj73ve+rFu3rtv2UaNGbbePYcOG7fLxYSCxQoc9YOjQoZk7d26S5M9//nOvHWfY\
sGHb/Vw6yXbhfLv169fnvvvuy+WXX55Zs2bl9NNPz8SJEzNhwoS8/PLLuzzPqFGjsnXr1qxZs6bb\
9mazmX/+858ZPXr0Lu8b6E7QoYetXr16h9vfPG09duzYXjt2R0dHnnjiiW7bFi9enI0bN+70dY1G\
I81mM8OGDeu2/fbbb8+2bdu6bXvzOe9m1Xz66acnSe66665u2++999689tprXV8Hdp9T7tDDzjzz\
zLz//e/P1KlTc8QRR+SNN97I448/nuuvvz4jR47MJZdc0mvHnjZtWubMmZNvfOMbmTRpUlasWJHv\
fOc72W+//Xb6un333Tcf//jHc91112X06NHp6OjIb37zm9xxxx3Zf//9uz336KOPTpLcdttt2Wef\
fTJ8+PCMGzduh6fLP/GJT+TMM8/MzJkz8+qrr+aUU07JE088kblz5+bDH/5wpk2b1mPvHQY6K3To\
YVdccUUOOOCA3HjjjTn77LPzyU9+MjfffHPOOOOM/OEPf8iECRN67diXXXZZLrvssixYsCBTp07N\
vffemx/96EfbRXlH7r777px66qm5/PLL89nPfjaPPvpofv3rX2/3n4Fx48Zl/vz5Wb58eSZPnpyJ\
EyfmF7/4xQ732Wg08rOf/SzTp0/P9773vUyZMqXrV9gWL1683RkBYNc1ms1ms91DAAC7xwodAAoQ\
dAAoQNABoABBB4ACBB0AChB0AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAoQNAB\
oABBB4ACBB0AChB0AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAoQNABoABBB4AC\
BB0AChB0AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAoQNABoABBB4ACBB0AChB0\
AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAoQNABoABBB4ACBB0AChB0AChA0AGg\
AEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAoQNABoABBB4ACBB0AChB0AChA0AGgAEEHgAIE\
HQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAoQNABoABBB4ACBB0AChB0AChA0AGgAEEHgAIEHQAKEHQA\
KEDQAaAAQQeAAgQdAAoQdAAoQNABoABBB4ACBB0AChB0AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAA\
QQeAAgQdAAoQdAAoQNABoABBB4ACBB0AChB0AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQd\
AAoQdAAoQNABoABBB4ACBB0AChB0AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAo\
QNABoABBB4ACBB0AChB0AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAoQNABoABB\
B4ACBB0AChB0AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAoQNABoABBB4ACBB0A\
ChB0AChA0AGgAEEHgAIEHQAKEHQAKEDQAaAAQQeAAgQdAAoQdAAoQNABoABBB4ACBB0AChB0AChA\
0AGgAEEHgAIEHQAKEHQAKEDQAaAAQYeB6vXXkx/+sPUn0O8JOgxU11+fnH9+csMN7Z4E6AGNZrPZ\
bPcQwB724ovJYYclmzcnI0YkTz+djB3b7qmA3WCFDgPRrFnJli2tv7/+ejJ7dnvnAXabFToMNMuW\
JSeeuOPtH/nInp8H6BGCDgPJG28kH/1o8sc/Jtu2vbV98ODk+ONbUW802jcfsMuccoeB5O67k0cf\
7R7zpPX4kUdaXwf6JSt0GCg2bmxdCLdmTbKjb/tGIxkzJnnmmWTkyD0/H7BbrNBhoLj22mTt2h3H\
PGltX7s2mTdvz84F9AgrdBgInnsuGT/+rSvbd2bo0GTlyqSjo9fHAnqOFToMBF/96juvzN+u2Ww9\
H+hXrNChuiVLklNP3bXXTZrU09MAvUTQobJt25Jjjkn++tftr2zfmcGDkyOOSJYvb/0d6POccofK\
br89WbHivcU8aT3/L39J7rijd+YCepwVOlT1yivJBz6Q/Pvfu76PAw5Inn022X//npoK6CVW6FDV\
N7+ZrF+/e/tYvz751rd6Zh6gV1mhQ0VPPZUcffR7P9W+I4MHt06/jx+/+/sCeo0VOlR06aU9d0/2\
RqO1P6BPE3So5pe/TBYuTLZu7Zn9bd2aPPBAa79An+WUO1Ty+uvJkUcmq1a1PlmtpwwalIwblzz5\
ZOtOckCfY4UOldxyS+s2rz0Z86S1v2efbe0f6JOs0KGKl15qfZraxo29d4yRI1thHzOm944B7BIr\
dKjiiiuSTZt69xibNrWOA/Q5VuhQweOPJ8cf/+4/gGV3NBrJY48lxx3X+8cC3jUrdOjvms3kS1/a\
c/dcHzw4+fKX98x/HoB3TdChv7v33uThh3vu19T+l61bk9//PvnJT/bM8YB3xSl36M82bUo+9KFk\
9eqev7J9ZwYNSsaOTVauTEaM2HPHBd6RFTr0Z9dfn7z44p6NedI63j/+kdxww549LvCOrNChv/r7\
31ur882b2zfD8OHJ008nBx/cvhmAJFbo0H/NnJls2dLeGbZsac0BtJ0VOvRHS5cmJ5/c7inesnRp\
cuKJ7Z4CBjRBh/7mjTeSiROT5ct75uNRd9fgwcmxxyaPPNK6WA5oC9990N/84AetG7v0hZgnrTke\
eyy56652TwIDmhU69CcbNiQf+ECybl3furFLo5GMHp0880yyzz7tngYGJCt06E+uvjp5+eW+FfOk\
Nc+6dck117R7EhiwrNChv3j22WT8+D13R7hdMXRo8tRTrbMIwB5lhQ79xfTp7Z7gf2s2kxkz2j0F\
DEhW6NAfLFqUnHFGu6d49xYtSk47rd1TwIAi6NDXbd2aTJiQ/O1vfefK9p0ZPDg5/PDkiSeSIUPa\
PQ0MGE65Q193222tn0v3h5gnrTmffDL57nfbPQkMKFbo0Je9/HLrArP169s9yXu3//6tC/kOOKDd\
k8CAYIUOfdlVVyUbN7Z7il2zYUNrfmCPsEKHvmrFitbPzvf0R6P2pEGDkj/9KTnqqHZPAuVZoUNf\
1GwmF1/c/++NPmhQcsklfe9GOFBQP//XAoq6777Wr3715ZvIvBtbtyYPPpjcf3+7J4HynHKHvub1\
11t3hHvhhf59uv1NgwYlhx7aulJ/r73aPQ2UZYUOfc3NNyfPP18j5knrfaxalfzf/7V7EijNCh36\
kn/9q/Vrav/5T7sn6Xl77936NLaDDmr3JFCSFTr0JV/7WrJ5c7un6B2bNydf/3q7p4CyrNChr9iw\
Idlvv9pXhDcarZvk+Mx06HFutAx9xciRrSvC16xp9yS9Z8yY1vsEepwVOgAU4GfoAFCAoANAAYIO\
AAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAU\
IOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICg\
A0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4A\
BQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg\
6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKAD\
QAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAF\
CDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDo\
AFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANA\
AYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUI\
OgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgA\
UICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0AB\
gg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6\
ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQ\
gKADQAGCDgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGC\
DgAFCDoAFCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoA\
FCDoAFCAoANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFCDoAFCA\
oANAAYIOAAUIOgAUIOgAUICgA0ABgg4ABQg6ABQg6ABQgKADQAGCDgAFCDoAFPD/y8ns7RuBaEcA\
AAAASUVORK5CYII=\
"


    /* set a timeout to make sure all the above elements are created before
       the object is initialized. */
    setTimeout(function() {
        anim3b9ce7f55fc64075901dd4cee46c8c5b = new Animation(frames, img_id, slider_id, 299.0,
                                 loop_select_id);
    }, 0);
  })()
</script>
